ENH: Teach rerun how to handle merges

kyleam opened the pull request:
Tagging as WIP because I want to test it out a bit more.

~- [ ] how should we handle cases where a merge is in the revision range but only one of the parents is?~
~- [ ] handle merges of unrelated branches~

Update: Lots to do in a reroll.  See https://github.com/datalad/datalad/pull/2754#issuecomment-412226197

yarikoptic commented on datalad/interface/rerun.py:
```diff
@@ -253,6 +253,12 @@ def _revs_as_results(dset, revs):
                 "Error on {}'s message: {}".format(rev, exc_str(exc)))
 
         if info is not None:
+            if len(parents) > 1:
```
Just curious if you did git this situation in some  case? Eg if some settings manage to suck in the full run record into the merge message

yarikoptic replied:
Cool, thanks

yarikoptic commented on datalad/interface/tests/test_run.py:
```diff
@@ -476,6 +468,84 @@ def test_rerun_cherry_pick(path):
+        assert_in("has run information but is a merge commit", cml.out)
+    eq_(expected_hexsha, ds.repo.get_hexsha())
+
+
```
Could add a test/support for merging branches without a common history? Eg if we did run in two branches without a common room and then merged allowing disconnected histories

kyleam replied:
This is one of the things I want to think about/play around with more.

yarikoptic commented on datalad/interface/rerun.py:
```diff
@@ -360,9 +372,41 @@ def _rerun(dset, results):
+                res["commit"], "--no-patch", "--format=%B")
+            dset.repo._git_custom_command(
+                None,
+                ["git", "merge", "--no-ff", "-m", msg] + new_parents[1:],
```
I wonder if it makes sense right away to reserve some options to allow to pass into merge eg to instruct on conflicts resolution strategy

kyleam replied:
A similar comment could be made about cherry picking non-run commits.  I don't plan to do that until the need arises and a convincing case is made.

yarikoptic commented on datalad/interface/rerun.py:
```diff
@@ -360,9 +372,41 @@ def _rerun(dset, results):
+            # We uniquify because content may not change (depending on the
+            # value of --onto).
+            new_parents = unique(
+                filter(None, (new_bases.get(p) for p in res["parents"])))
```
Ha, didn't know about None-- would have used `bool` here. Good to know I guess

kyleam replied:
Why do you guess?

yarikoptic replied:
Because this is not a special behavior which isn't as clear (requires reading the function documentation) as just using `bool` as the comparison/casting function here

kyleam replied:
Agreed.  Will change.

yarikoptic commented on datalad/interface/rerun.py:
```diff
@@ -360,9 +372,41 @@ def _rerun(dset, results):
+                lgr.warning(
+                    "Skipping merge %s because only one parent "
+                    "is in the specified revision range",
+                    res["commit"])
```
Hm, if I was to specify a range for rerunning - don't we still want to merge previous histories (as is, without rerunning) if that was done on original history?

kyleam replied:
Good question.  Maybe?  Another thing I need to think about more.

kyleam commented:
Time conservation notice: These are my notes from exploring (shortcomings of) the current implementation. You probably don't want to bother reading anything besides the summary below.

---

- [Summary of problems with current implementation](#org6387771)
- [Replay cases](#org0de3869)
  - [No commits on left, run commit on right](#org4862883)
  - [Run commits left and right](#orgc776eb5)
  - [Non-run commit on left then run commit right](#orgf278365)
  - [Run commit on left then non-run commit on right](#orge0ee1e5)
  - [Non-run commits on left and right](#org398fa95)
  - [Unrelated histories](#orgaf819aa)
    - [Run commit on left then non-run commit on right](#org83e7836)
    - [Non-run commit on left then run commit on right](#org9b005ca)
- [Parents outside of the revision range](#org7eb75f5)
- [Reun on top](#orge4315bc)
- [Octopus merges](#orgfcf2eea)
- [Merge conflicts](#org9285c17)



<a id="org6387771"></a>

### Summary of problems with current implementation

-   We need to consider the base when cherry picking.

-   We drop leading commits.

    This is sensible in the non-merge case, but we can't do it unconditionally in the merge case.

-   We don't detect unrelated histories. Instead they get recreated as though they forked from a common ancestor.

-   If a merge is included in the revision range but a parent isn't, we should substitute the parent as is when we merge the reran parents.

-   If we're rerunning commits on top of themselves, we don't find the proper base when rerunning commits on two different lines results in changes.

---

In the diagrams below, the following notation is used:

-   x\_n: commit x without a run command
-   x\_r: commit x with a run command
-   x\_C: commit x\_n was cherry picked on rerun
-   x\_M: merge commit x\_n was remerged on rerun
-   x\_R: run commit x\_r was reran

The commits in the diagrams are ordered by the commit date (i.e. how they would be ordered by `git rev-list` by default and unlike the default `--topo-order` implied by `git log --graph`.)


<a id="org0de3869"></a>

### Replay cases

Below are several cases of replaying a sequence of commits on a common base with `datalad rerun --since= --onto=`. This isn't an exhaustive list of cases, but I've tried to boil it down to ones that show different core problems of the current implementation. They're labeled with "[OK]" if the current implementation handles them "correctly" and "[BAD]" otherwise.


<a id="org4862883"></a>

#### [OK] No commits on left, run commit on right


    o                 c_n
    |\
    | o               b_r
    |/
    o                 a_n

`rerun` recreates this fine:

    o                 c_M
    |\
    | o               b_R
    |/
    o                 a_n

Also recreated OK if there are stem run commits.


<a id="orgc776eb5"></a>

#### [OK] Run commits left and right

Each side has a run commit.

    o                 d_n
    |\
    o |               c_r
    | o               b_r
    |/
    o                 a_n

Result:

    o                 d_M
    |\
    o |               c_R
    | o               b_R
    |/
    o                 a_n


<a id="orgf278365"></a>

#### [BAD] Non-run commit on left then run commit right

"left" and "right" don't matter here. What is important is that the non-run commit is encountered first when walking the history with `git rev-list --reverse`.

    o                 d_n
    |\
    | o               c_r
    o |               b_n
    |/
    o                 a_n

Rerunning this gives

    o                 c_R
    o                 a_n

We want a merge, but `b_n` is dropped because of the "drop non-run stem commits" step. If there were a rerun commit before the fork, the merge would be replayed fine.


<a id="orge0ee1e5"></a>

#### [BAD] Run commit on left then non-run commit on right

Again, "left" and "right" don't matter. What does it that we encounter the run commit before the other line's non-run commit.

    o                 d_n
    |\
    o |               c_n
    | o               b_r
    |/
    o                 a_n

Result:

    o                 c_C
    o                 b_R
    o                 a_n

We want a merge. Cherry pick needs to consider what its base should be (`b_R`). Unlike the previous case, stem run commits wouldn't help here.


<a id="org398fa95"></a>

#### [BAD] Non-run commits on left and right

    o                 e_n
    |\
    o |               d_n
    | o               c_n
    |/
    o                 b_r
    o                 a_n

Result:

    o                 d_C
    o                 c_C
    o                 b_R
    o                 a_n

This fails for the same reason as the previous one. We should identify the base for cherry pickng as `b_R`.


<a id="orgaf819aa"></a>

#### Unrelated histories

It'd be convenient if we could just barf if we hit orphan branches. Unfortunately, the crawler uses them, so it isn't unlikely that we'll be in a repo that has them.


<a id="org83e7836"></a>

##### [BAD] Run commit on left then non-run commit on right

    o                 d_n
    |\
    | o               c_n
    o                 b_r
    o                 a_n

Result:

    o                 d_M
    |\
    | o               c_C
    |/
    o                 b_R
    o                 a_n

We need to recognize that the histories are unrelated and not fork `c_C` from `b_R`.


<a id="org9b005ca"></a>

##### [BAD] Non-run commit on left then run commit on right

    o                 e_n
    |\
    | o               d_r
    | o               c_n
    o                 b_n
    o                 a_n

Result:

    o                 d_R
    o                 c_n

We need to avoid unconditionally dropping stem non-run commits.


<a id="org7eb75f5"></a>

### Parents outside of the revision range

It's possible to specify a revision range to `rerun` that doesn't include all parents of the merge.

There are two possible behaviors in this case:

-   Don't do the merge (or in the case of an octopus merge, drop the excluded parents from the merge).

-   Merge in the original parent.

Looking at the simplest case

    o                 c_n
    |\
    | o               b_r
    |/
    o                 a_n

say we call `datalad rerun --onto= --since=b_r c_n`. That translates to running `b_r..c_c_n` (i.e. `[a_n, c_n]`). Neither of those commits are run commits, so the only option is to report that no run commands were found.

The next simplest case is

    o                 d_n
    |\
    | o               c_r
    o |               b_r
    |/
    o                 a_n

Here `c_n..d_n` would include a run commit, `b_r`. Going with the first behavior (dropping the `c_r` line) seems like it'd be a mistake because a commit after `d_n` may depend on something in the history of `c_r`. The second behavior (merging in `c_r`) avoids this. Also, doing the merge doesn't seem to go against the revision range restriction, since that can be viewed as a specification of which commits to *rerun*, not which commits to include in the final history.

In diagrams, favor

    o                 d_M
    |\
    | o               c_r
    o |               b_R
    |/
    o                 a_n

over

    o                 b_R
    o                 a_n


<a id="orge4315bc"></a>

### Rerun on top

Running commits on top of each other isn't as straightforward conceptually as `--onto=''`. We can't expect the history to always be recreated because, if rerunning the commit doesn't change the content, there's nothing to commit.

    o                 d_c
    |\
    | o               c_r
    o |               b_r
    |/
    o                 a_n

If we run `datalad rerun --since=` and both `b_r` and `c_r` don't have any changes, we'll remain on `d_c` with no changes. Trying to merge `c_r` would be a noop. (The current implementation gives a misleading message, but is otherwise OK here.)

Instead, if `c_r` has a change on rerun, the current implementation ends up with

    o                 d_M
    |\
    | o               c_R
    |/
    o                 d_c
    |\
    | o               c_r
    o |               b_r
    |/
    o                 a_n

That seems like the most reasonable thing to do. Likewise, if just `b_r` changes on rerun, we get something reasonable:

    o                 b_R
    o                 d_c
    |\
    | o               c_r
    o |               b_r
    |/
    o                 a_n

However, if the results from both `b_r` and `c_r` change on rerun, we do the wrong thing:

    o                 d_M
    |\
    | o               c_R
    |/
    o                 b_R
    o                 d_c
    |\
    | o               c_r
    o |               b_r
    |/
    o                 a_n

Notice that the fork happens at the wrong base.


<a id="orgfcf2eea"></a>

### Octopus merges

Octopus merges have the same issues as standard merges, but I don't see any obvious additional issues. If they end up being a headache, I plan to drop support for them. But, at the moment, they don't seem to add much complexity.


<a id="org9285c17"></a>

### Merge conflicts

If a rerun hits a merge conflict, it will fail (in the same way that a rerun already fails if there's conflict in a cherry pick). We should be able to address this by dumping the current state (the remaining results list) to .git/datalad-rerun-merge or similar and then exiting. Once the user has resolved the merge, they could signal to continue with `datalad rerun --continue` ([related issue](https://github.com/datalad/datalad/issues/2092)).

yarikoptic commented:
Kudos @kyleam

kyleam commented:
I've pushed an update.  Things are at a state where I'd appreciate people trying it out and letting me know what they think.

The Travis runs are failing across the board.  I haven't looked at it in detail, though I'll boldly claim it is unrelated.

mih commented:
Could you please address the conflicts? Thx.

kyleam commented:
@mih:

> I had a walk through the code and tests and it looks good to me.

Thanks for taking a look!

> Thx for anticipating this need, it will likely be in the spotlight of our activities a few months from now.

Interesting.  Looking forward to seeing (or at least hearing) about what that involves.

> I am afraid that until then my concept of what we would need to be able to support will remain somewhat limited and theoretical. From this POV, however, this PR is exactly what is needed ;-)

Yeah, I'm of course working from the same point :/  I was hoping (though not really expecting) that a user would jump in and say "hey, great, I need this".

> The only criticism I have is that in the light of the movement away from GitPy, the tests should rather not introduce additional usage of the GitPy API.

Thanks for catching that.  I cleaned up some of those in previous iterations, but I see that there are a few spots I missed.  I'll deal with those when I rebase to resolve the recently emerged conflicts.

kyleam commented:
>  I'll deal with those when I rebase to resolve the recently emerged conflicts.

Done.

<details>
<summary>range-diff</summary>

```
 1:  76b25a9bc !  1:  ebe8d7f2b RF: rerun: Move intermediate variable into helper function
    @@ -15,7 +15,7 @@
      
     -def _revs_as_results(dset, revs):
     +def _revrange_as_results(dset, revrange):
    -+    revs = dset.repo.repo.git.rev_list("--reverse", revrange, "--").split()
    ++    revs = dset.repo.get_revisions(revrange, options=["--reverse"])
          for rev in revs:
              res = get_status_dict("run", ds=dset, commit=rev)
              full_msg = dset.repo.format_commit("%B", rev)
    @@ -23,7 +23,7 @@
          In the standard case, the information in these results will be used to
          actually re-execute the commands.
          """
    --    revs = dset.repo.repo.git.rev_list("--reverse", revrange, "--").split()
    +-    revs = dset.repo.get_revisions(revrange, options=["--reverse"])
     +
          try:
     -        results = _revs_as_results(dset, revs)
 2:  20c01d444 !  2:  8433edf3d RF: rerun: Add parents to the revision results
    @@ -12,15 +12,19 @@
      
      
      def _revrange_as_results(dset, revrange):
    --    revs = dset.repo.repo.git.rev_list("--reverse", revrange, "--").split()
    +-    revs = dset.repo.get_revisions(revrange, options=["--reverse"])
     -    for rev in revs:
     -        res = get_status_dict("run", ds=dset, commit=rev)
    -+    out = dset.repo.repo.git.rev_list("--reverse", "--parents", revrange, "--")
    -+    if not out:
    ++    rev_lines = dset.repo.get_revisions(
    ++        revrange, fmt="%H %P", options=["--reverse"])
    ++    if not rev_lines:
     +        return
     +
    -+    for rev_line in out.splitlines():
    -+        fields = rev_line.split(" ")
    ++    for rev_line in rev_lines:
    ++        # The strip() below is necessary because, with the format above, a
    ++        # commit without any parent has a trailing space. (We could also use a
    ++        # custom `rev-list --parents ...` call to avoid this.)
    ++        fields = rev_line.strip().split(" ")
     +        rev, parents = fields[0], fields[1:]
     +        res = get_status_dict("run", ds=dset, commit=rev, parents=parents)
              full_msg = dset.repo.format_commit("%B", rev)
 3:  438483951 !  3:  0a9250f64 RF: rerun: List revisions with --topo-order
    @@ -13,11 +13,10 @@
      +++ b/datalad/interface/rerun.py
     @@
      
    - 
      def _revrange_as_results(dset, revrange):
    --    out = dset.repo.repo.git.rev_list("--reverse", "--parents", revrange, "--")
    -+    out = dset.repo.repo.git.rev_list("--reverse", "--parents", "--topo-order",
    -+                                      revrange, "--")
    -     if not out:
    +     rev_lines = dset.repo.get_revisions(
    +-        revrange, fmt="%H %P", options=["--reverse"])
    ++        revrange, fmt="%H %P", options=["--reverse", "--topo-order"])
    +     if not rev_lines:
              return
      
 4:  6faaa305c =  4:  0f3a05a52 RF: rerun: Ignore multi-parent run commits
 5:  9692e2551 !  5:  32e5abd13 RF: rerun: Refuse to run root commits
    @@ -55,17 +55,17 @@
     -    # Check out an orphan branch so that we can test the "one commit
     -    # in a repo" case.
          ds.repo.checkout("orph", options=["--orphan"])
    -     ds.repo.repo.git.reset("--hard")
    +     ds.repo._git_custom_command(None, ["git", "reset", "--hard"])
          ds.repo.config.reload()
      
          ds.run('echo static-content > static')
    -     assert_result_count(ds.repo.repo.git.rev_list("HEAD").split(), 1)
    +     eq_(len(ds.repo.get_revisions("HEAD")), 1)
     -
     -    # Rerunning with just one commit doesn't raise an error ...
     -    ds.rerun()
     -    # ... but we're still at one commit because the content didn't
     -    # change.
    --    assert_result_count(ds.repo.repo.git.rev_list("HEAD").split(), 1)
    +-    eq_(len(ds.repo.get_revisions("HEAD")), 1)
     -
     -    # We abort rather than trying to do anything when --onto='' and
     -    # --since='' are given together and the first commit contains a
 6:  8d4855d7c =  6:  709e58677 DOC: rerun: Redefine --onto=''
 7:  0b07c2358 =  7:  e975b5639 RF: rerun: Unconditionally drop stem non-run commits
 8:  b99f6c7c3 =  8:  e0212d7f8 RF: rerun: Resolve start point in "checkout" result
 9:  a6c0e232a =  9:  abfb98411 RF: rerun: Rearrange a conditional
10:  9306b3a7f = 10:  57dce8be4 RF: rerun: Restructure an if-elif chain
11:  6b6cb6ee1 = 11:  274200267 RF: rerun: Pull out the result's hexsha into a variable sooner
12:  e4a590307 = 12:  26f091f2c RF: rerun: Rename a variable
13:  6d6192598 = 13:  7fbc6e460 RF: rerun: Use a more general approach to decide "skip" or "pick"
14:  6e0b91e79 ! 14:  a276c6410 ENH: rerun: Support rerunning history with merges
    @@ -29,7 +29,7 @@
              else:
                  revrange = "{}..{}".format(since, revision)
      
    --        if ds.repo.repo.git.rev_list("--merges", revrange, "--"):
    +-        if ds.repo.get_revisions(revrange, options=["--merges"]):
     -            yield get_status_dict(
     -                "run", ds=ds, status="error",
     -                message="cannot rerun history with merge commits")
    @@ -158,7 +158,7 @@
     +    if branch_to_restore:
     +        # The user asked us to replay the sequence onto a branch, but the
     +        # history had merges, so we're in a detached state.
    -+        dset.repo.repo.git.update_ref("refs/heads/" + branch_to_restore,
    ++        dset.repo.update_ref("refs/heads/" + branch_to_restore,
     +                             "HEAD")
     +        dset.repo.checkout(branch_to_restore)
      
    @@ -194,27 +194,25 @@
     +    ds = Dataset(path).create()
     +    ds.run("echo foo >>foo")
     +    ds.run("echo invalid >>invalid")
    -+    run_msg = ds.repo.repo.git.show(
    -+        "--no-patch", "--format=%B", "HEAD")
    ++    run_msg = ds.repo.format_commit("%B")
     +    run_hexsha = ds.repo.get_hexsha()
    -+    ds.repo.repo.git.reset("--hard", "master~")
    ++    ds.repo._git_custom_command(None, ["git", "reset", "--hard", "master~"])
     +    with open(op.join(ds.path, "non-run"), "w") as nrfh:
     +        nrfh.write("non-run")
     +    ds.save()
     +    # Assign two parents to the invalid run commit.
    -+    commit = ds.repo.repo.git.commit_tree(
    -+        run_hexsha + "^{tree}",
    -+        "-m", run_msg,
    ++    commit = ds.repo._git_custom_command(
    ++        None,
    ++        ["git", "commit-tree", run_hexsha + "^{tree}", "-m", run_msg,
     +         "-p", run_hexsha + "^",
    -+        "-p", ds.repo.get_hexsha())
    -+    ds.repo.repo.git.reset("--hard", commit)
    ++         "-p", ds.repo.get_hexsha()])[0].strip()
    ++
    ++    ds.repo._git_custom_command(None, ["git", "reset", "--hard", commit])
     +    hexsha_orig = ds.repo.get_hexsha()
     +    with swallow_logs(new_level=logging.WARN) as cml:
     +        ds.rerun(since="")
     +        assert_in("has run information but is a merge commit", cml.out)
    -+    assert_result_count(
    -+        ds.repo.repo.git.rev_list(hexsha_orig + "..master").split(),
    -+        1)
    ++    eq_(len(ds.repo.get_revisions(hexsha_orig + "..master")), 1)
     +
     +
      @known_failure_windows
```

</details>

mih commented:
Good to go from my POV. As I said, this looks like a good thing to have and likely needs to be revisited once speculation about features can be replaced empirical evidence.

kyleam commented:
> Good to go from my POV.

Thanks.  I'll plan to merge this after 0.12.0rc5 unless others raise objections in the meantime.
